core/src/main/kotlin/entity/Icon.kt

```diff
@@ -0,0 +1,75 @@
+package dev.kord.core.entity
+
+import dev.kord.core.Kord
+import dev.kord.core.KordObject
+import dev.kord.core.cache.data.EmojiData
+import dev.kord.core.cache.data.MemberData
+import dev.kord.core.cache.data.UserData
+import dev.kord.rest.Image
+
+sealed class Icon(override val kord: Kord, val animated: Boolean, private val rawAssetUri: String) : KordObject {
+
+    val format: Image.Format
+        get() = when {
+            animated -> Image.Format.GIF
+            else -> Image.Format.WEBP
+        }
+
+    val url: String
+        get() = "$rawAssetUri.${format.extension}"
+
+    fun getUrl(format: Image.Format): String? {
+        if (format == Image.Format.GIF && !animated) return null
+        return "$rawAssetUri.${format.extension}"
+    }
+
+    fun getUrl(size: Image.Size): String {
+        return "$url?size=${size.maxRes}"
+    }
+
+    fun getUrl(format: Image.Format, size: Image.Size): String? {
+        if (format == Image.Format.GIF && !animated) return null
+        return "$rawAssetUri.${format.extension}?size=${size.maxRes}"
+    }
+
+    suspend fun getImage(): Image = Image.fromUrl(kord.resources.httpClient, url)
+
+    suspend fun getImage(size: Image.Size): Image = Image.fromUrl(kord.resources.httpClient, getUrl(size))
+
+    suspend fun getImage(format: Image.Format): Image? =
+        getUrl(format)?.let { Image.fromUrl(kord.resources.httpClient, it) }
+
+    suspend fun getImage(format: Image.Format, size: Image.Size): Image? =
+        getUrl(format, size)?.let { Image.fromUrl(kord.resources.httpClient, it) }
+
+    override fun toString(): String {
+        return "Icon(type=${javaClass.name},animated=$animated,rawAssetUri=$rawAssetUri,kord=$kord)"
+    }
+
+    class EmojiIcon(kord: Kord, data: EmojiData) : Icon(
+        kord,
+        data.animated.discordBoolean,
+        "$CDN_BASE_URL/emojis/${data.id.asString}"
```

Comment:
make the rawAssertUri an abstract property and not a part of the constructor so you'd be able to define it inside the entity itself. Maybe create a `Route` like implementation for these.

Reply:
The inner classes of the Icon class, are kind of a "Route" enum. I really like this approach and I think you should reconsider this.
The rawAssetUri is a private val on purpose, since it does not contain a format yet and is used in the `val url` field which appends the default format.

Would you want to reconsider this, or is this decision final?
Please let me know, so I can adjust if needed.

Reply:
None of the feedback I gave is final.
I'm not speaking about making them inner classes. I don't see a problem with that. tho I don't like that you are explicitly passing the string there.
Maybe if you could make it so someone else can do something like `CDN.Emojis` would be of great help for everyone :D